File: backend/app/schemas/experiments.py
"""Typed Stage 10 experiment contracts.

Format-level constraints (lengths, required fields) live here; the product rules
(date order, duration, edit policy, text normalisation) live in
``app.domain.experiments`` so a single place decides what a valid experiment is.
"""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field, model_validator

from app.db.models import Experiment
from app.domain.experiments import (
    TEXT_MAX_LENGTH,
    TITLE_MAX_LENGTH,
    ExperimentAnalysis,
    ExperimentPhase,
    ExperimentStatus,
    phase_for,
    status_of,
)
from app.domain.owl import OwlState


class ExperimentCreate(BaseModel):
    title: str = Field(min_length=1, max_length=TITLE_MAX_LENGTH)
    hypothesis: str = Field(min_length=1, max_length=TEXT_MAX_LENGTH)
    protocol: str = Field(min_length=1, max_length=TEXT_MAX_LENGTH)
    start_date: date
    end_date: date


class ExperimentUpdate(BaseModel):
    """Partial update: send only the fields that change."""

    title: str | None = Field(default=None, min_length=1, max_length=TITLE_MAX_LENGTH)
    hypothesis: str | None = Field(default=None, min_length=1, max_length=TEXT_MAX_LENGTH)
    protocol: str | None = Field(default=None, min_length=1, max_length=TEXT_MAX_LENGTH)
    start_date: date | None = None
    end_date: date | None = None

    @model_validator(mode="after")
    def _require_something_to_change(self) -> ExperimentUpdate:
        if all(value is None for value in (self.title, self.hypothesis, self.protocol,
                                           self.start_date, self.end_date)):
            raise ValueError("Provide at least one field to update.")
        return self


class ExperimentRead(BaseModel):
    id: int
    title: str
    hypothesis: str
    protocol: str
    start_date: date
    end_date: date
    status: ExperimentStatus
    phase: ExperimentPhase
    cancelled_on: date | None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_model(cls, experiment: Experiment, *, today: date) -> ExperimentRead:
        cancelled_on = experiment.cancelled_on
        return cls(
            id=experiment.id,
            title=experiment.title,
            hypothesis=experiment.hypothesis,
            protocol=experiment.protocol,
            start_date=experiment.start_date,
            end_date=experiment.end_date,
            status=status_of(experiment.start_date, experiment.end_date, cancelled_on, today),
            phase=phase_for(experiment.start_date, experiment.end_date, cancelled_on, today),
            cancelled_on=cancelled_on,
            created_at=experiment.created_at,
            updated_at=experiment.updated_at,
        )


class ExperimentOverlapRead(BaseModel):
    id: int
    title: str
    start_date: date
    end_date: date


class ExperimentListRead(BaseModel):
    experiments: list[ExperimentRead]
    # Stage 9 Owl reuse: a single experiment-related banner for the list page.
    owl: OwlState | None = None


class ExperimentDetailRead(BaseModel):
    experiment: ExperimentRead
    overlaps: list[ExperimentOverlapRead]
    # Descriptive before/during/after comparison over the canonical dataset.
    analysis: ExperimentAnalysis

File: frontend/src/api/experiments.ts
/**
 * Stage 10 experiment endpoints.
 *
 * The frontend never computes a comparison: the backend describes the
 * before/during/after windows over the canonical dataset. This module only
 * shapes requests and mirrors the response contract.
 */

import { apiRequest } from './client'
import type { OwlState } from './owl'

export type ExperimentStatus = 'scheduled' | 'active' | 'completed' | 'cancelled'
export type ExperimentStage = 'before' | 'during' | 'after'

export interface ExperimentWindow {
  key: ExperimentStage
  start: string
  end: string
}

export interface ExperimentWindows {
  before: ExperimentWindow
  during: ExperimentWindow
  after: ExperimentWindow
}

export interface ExperimentPhase {
  status: ExperimentStatus
  stage: ExperimentStage
  day_index: number | null
  days_total: number | null
  days_until_start: number | null
  days_since_end: number | null
  after_collected_days: number
  after_total_days: number
}

export interface Experiment {
  id: number
  title: string
  hypothesis: string
  protocol: string
  start_date: string
  end_date: string
  status: ExperimentStatus
  phase: ExperimentPhase
  cancelled_on: string | null
  created_at: string
  updated_at: string
}

export interface PeriodCoverage {
  calendar_days: number
  elapsed_days: number
  observed_days: number
  /** 0..1, or null when the period has no calendar days at all. */
  coverage: number | null
}

export interface OverallPeriod {
  coverage: PeriodCoverage
  mean_score: number | null
  scored_days: number
  completed_weight: number
  required_weight: number
}

export interface OverallComparison {
  before: OverallPeriod
  during: OverallPeriod
  after: OverallPeriod
  delta_before_during: number | null
  delta_during_after: number | null
  delta_before_after: number | null
}

export interface HabitPeriod {
  obligation_days: number
  observed_days: number
  done_days: number
  missed_days: number
  completion_ratio: number | null
}

export interface HabitComparison {
  habit_id: number
  name: string
  before: HabitPeriod
  during: HabitPeriod
  after: HabitPeriod
  delta_during_vs_before: number | null
}

export interface StatePeriod {
  observed_days: number
  value: number | null
}

export interface StateComparison {
  key: string
  label: string
  kind: string
  before: StatePeriod
  during: StatePeriod
  after: StatePeriod
  delta_during_vs_before: number | null
}

export interface SeriesPoint {
  date: string
  phase: ExperimentStage
  score: number | null
  energy: number | null
}

export interface ExperimentAnalysis {
  windows: ExperimentWindows
  overall: OverallComparison
  habits: HabitComparison[]
  state: StateComparison[]
  series: SeriesPoint[]
  sufficient: boolean
  summary: string
}

export interface ExperimentOverlap {
  id: number
  title: string
  start_date: string
  end_date: string
}

export interface ExperimentListRead {
  experiments: Experiment[]
  owl: OwlState | null
}

export interface ExperimentDetailRead {
  experiment: Experiment
  overlaps: ExperimentOverlap[]
  analysis: ExperimentAnalysis
}

export interface ExperimentInput {
  title: string
  hypothesis: string
  protocol: string
  start_date: string
  end_date: string
}

/** Partial update: send only the fields that change. */
export type ExperimentUpdateInput = Partial<ExperimentInput>

export function fetchExperiments(signal?: AbortSignal): Promise<ExperimentListRead> {
  return apiRequest<ExperimentListRead>('/api/experiments', { signal })
}

export function fetchExperiment(id: number, signal?: AbortSignal): Promise<ExperimentDetailRead> {
  return apiRequest<ExperimentDetailRead>(`/api/experiments/${id}`, { signal })
}

export function createExperiment(input: ExperimentInput): Promise<Experiment> {
  return apiRequest<Experiment>('/api/experiments', { method: 'POST', body: input })
}

export function updateExperiment(id: number, input: ExperimentUpdateInput): Promise<Experiment> {
  return apiRequest<Experiment>(`/api/experiments/${id}`, { method: 'PATCH', body: input })
}

export function cancelExperiment(id: number): Promise<Experiment> {
  return apiRequest<Experiment>(`/api/experiments/${id}/cancel`, { method: 'POST' })
}
